### DISCUSSION TITLE
Unable to open Actual Budget app via ScriptKit

### DISCUSSION THREAD

**Discussion by JamesHenry** [johnlindquist/kit#314] [Errors] [answered]
As per our twitter discussion.

- https://actualbudget.com/ is an electron desktop app
- in order to interact with your local data via its API, the app needs to be open
- I have a scriptkit implementation of my API logic, but I cannot get scriptkit to open the app for me successfully

You can open the app via e.g. the VSCode terminal with:

```sh
open -a Actual.app
```

But if you create a minimal scriptkit script:

**open-actual.js**

```js
exec(`open -a "Actual.app"`); // bounces for a sec, doesn't open

// exec(`open -a "Spotify.app"`); // other apps I have tried seem to work fine...
```

**Comment by jlongster** [comment]
Thanks for looking into this!

I realized that Actual might not be a "normal"  electron app. I'm so used to how it works that I forget it might be a little different.

One thing it does on startup is it forks the process which creates a background node process. This is the backend that runs the whole app. I don't know if that has any implications here, but that is one special thing it does compared to other Electron apps.

**Reply by johnlindquist** [reply]
Do you set any Environment variables when forking the background process?

**Reply by jlongster** [reply]
Not just for that process, but there are some env vars that the forked process inherits and are required to run:

Code for forking:

```
  serverProcess = fork(__dirname + '/server.js', [
    '--subprocess',
    app.getVersion(),
    socketName
  ]);
```

These two env vars get set in the electron main process and get inherited by the forked process:

```
  process.env.ACTUAL_DOCUMENT_DIR = app.getPath('documents');
  process.env.ACTUAL_DATA_DIR = app.getPath('userData');
```

I can take a look at some point as well, but won't have time for a little bit

**Comment by johnlindquist** [comment, marked as answer]
@JamesHenry looks like clearing the [ELECTRON_RUN_AS_NODE env var](https://www.electronjs.org/docs/api/environment-variables#electron_run_as_node) fixes it:

```js
delete process.env.ELECTRON_RUN_AS_NODE

exec(`open -a "Actual.app"`)
```

Script Kit works on the same "forking processes" principle (each script is a forked process controlled by the main Kit.app) that @jlongster uses for forking a background process. Script Kit's `fork` call is inheriting that env var from Electron which then conflicts with launching an Electron app which forks another node process (_PHEW_ 😅)

I _think_ I can safely delete that env var from my `fork` call to avoid anything like this in the future. I'll run some tests.

Thanks for the help @jlongster

**Reply by jlongster** [reply]
Cool, thanks @johnlindquist!

In the future our API will be a simple node script and not require Actual to be running, which should simplify JamesHenry's use case.

**Reply by JamesHenry** [reply]
Awesome, thanks for looking into this, Gents!

I’m AFK for a few hours but look forward to trying this out later
